core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -108,6 +110,51 @@ class TransactionCoordinator(brokerId: Int,
     }
   }
 
+  def handleAddPartitionsToTransaction(transactionalId: String,
+                                       pid: Long,
+                                       epoch: Short,
+                                       partitions: collection.Set[TopicPartition],
+                                       responseCallback: AddPartitionsToTxnCallback): Unit = {
+    if (transactionalId == null || transactionalId.isEmpty) {
+      // first check that transactional id is specified
+      responseCallback(Errors.INVALID_REQUEST)
+    } else if (!txnManager.isCoordinatorFor(transactionalId)) {
+      // check if it is the assigned coordinator for the transactional id
+      responseCallback(Errors.NOT_COORDINATOR)
+    } else if (!txnManager.isCoordinatorLoadingInProgress(transactionalId)) {
+      // check if the corresponding partition is still being loaded
+      responseCallback(Errors.COORDINATOR_LOAD_IN_PROGRESS)
+    } else {
+      // only try to get a new pid and update the cache if the transactional id is unknown
+      txnManager.getTransaction(transactionalId) match {
+        case None =>
+          responseCallback(Errors.INVALID_PID_MAPPING)
```

Comment:
Hmm. the comment on line 128 says ` only try to get a new pid and update the cache if the transactional id is unknown`, but this seems to just give up. It should be an error if there is no transaction metadata for the pid at this stage, so I think the comment should be corrected.